www/app/blog/2024-11-19-distributed-ddl.mdx

```diff
@@ -0,0 +1,230 @@
+
+So the `nile_ddl` extension takes the DDL and tries to determine which locks it will need. `ALTER` and `DROP` are fairly straight forward, we need to acquire a lock on the objects that the DDL is modifying. In most cases, acquiring the lock is simply a matter of sending `LOCK TABLE <table name> IN <lock level> MODE` with the same level that the DDL will require later. So before distributing `DROP TABLE todos`, the transaction coordinator must distribute `LOCK TABLE todos ACCESS EXLUSIVE MODE`. However, `LOCK TABLE` only applies to... well, tables. And we sometimes need to lock other objects. For example `DROP SCHEMA` requires locking the schema. Postgres has internal methods for locking other objects, so in our extension, we added the following logic:
+
+```c
+	if (is_rel)
+		LockRelationOid(objid, lockmode);
+	else
+		LockDatabaseObject(classid, objid, objsubid, lockmode);
+```
+
+If we need to lock a table (also called relation), we use `LockRelationOid`, the Postgres method behind `LOCK TABLE`. If we need to lock another object, we use `LockDatabaseObject` and directly lock the object we need.
+
+This is simple enough when acquiring locks locally, but how do we distribute this logic? We wrapped the object lock acquisition logic in the `nile_ddl` extension in a function and then exported it as a function:
+
+```c
+PGDLLEXPORT Datum nile_ddl_objectlock(PG_FUNCTION_ARGS);
+
+PG_FUNCTION_INFO_V1(nile_ddl_objectlock);
+```
+
+Then the transaction coordinator can send `SELECT nile_ddl.nile_ddl_objectlock(...)` to all the databases and acquire all the necessary locks for the DDL.
+
+When the transaction coordinator distributes the actual DDL, Postgres will try to re-acquire the locks. Because the DDL execution will happen in the same session in which we acquired the locks initially, and the locks are on the same objects and with the same level, we are guaranteed that the DDL processing will successfully acquire these locks. 
+
+So far we dealt with acquiring locks on existing tables and other objects. But there is one other case we need to handle before we are done. `CREATE TABLE` also requires locks - in order to avoid conflict between two sessions attempting to create the same table. How can we lock a table that does not exist yet? 
+
+Normally Postgres uses the data dictionary for this. It inserts a row for this new table into several tables in pg_catalog, the table name is added to unique indexes, and any attempt to insert another row with the same name into the same unique index will first block and then fail after the first session commits. Unfortunately, we can't use the same method. Row locks behave differently from table locks - they can't get re-acquired. Which means that if we manage to successfully acquire such lock, the actual DDL will then block on the same lock. 
```

Comment:
@rjuju can you double check that I didn't mis-state the reasons for using USERLOCK for `create table`? I didn't have great notes on why we can't use row locks...

Reply:
I added a few minor comments, otherwise it looks good thanks a lot!

Reply:
TY!